### ISSUE TITLE
meta-issue on C runtime API and compatibility stories

### ISSUE THREAD

**Issue by gasche** [ocaml/ocaml#9205]
There is a lot of discussion in preparation for 4.10 about the compatibility of programs linking to the C runtime, and to be honest I understand very little of it. I'm creating this issue in the hope to encourage a didactic discussion of the various choices, possibly clarify the design space and improve our interface, and encourage people to write proper documentation of the best practices.

Related issues, with their relevant keywords:
- #1798: CAML_API_VERSION
- #8713: `Caml_state` introduction and unplanned backward-compatibility disaster (tracking issue #9160), with follow-up PRs including #8940, #9115, #9143, #9202
- #9167: CAML_NAME_SPACE, CAML_INTERNALS and compatibility.h
- #9179: public/internal distinction and compatibility.h

Currently the only documentation of these matters in the OCaml manual is made of two paragraphs in http://caml.inria.fr/pub/docs/manual-ocaml/intfc.html#sec421 :

> By default, header files in the caml/ subdirectory give only access to the public interface of the OCaml runtime. It is possible to define the macro CAML_INTERNALS to get access to a lower-level interface, but this lower-level interface is more likely to change and break programs that use it.
> 
> Note: It is recommended to define the macro CAML_NAME_SPACE before including these header files. If you do not define it, the header files will also define short names (without the caml_ prefix) for most functions, which usually produce clashes with names defined by other C libraries that you might use. Including the header files without CAML_NAME_SPACE is only supported for backward compatibility.

This documentation is woefully insufficient to understand the current-times discussions, and most authors of affected OCaml software (in addition to a good share of the OCaml maintainers or contributors) are completely lost. As far as I know, none of the PRs above have considered extending or clarifying this documentation.

Could we try to clarify the design for the C runtime interface and compatibility mechanisms, have a proper documentation of this design, and implement it?

**Comment by gasche** [comment]
I would like to understand answers to the following questions:

- What is the responsibility of `compatibility.h`? (Who is in charge of the decision to include it in a C program?)
- How does a program signal that they use the internals of the OCaml runtime? What are our recommendation for upgrading those programs as the runtime evolves?
- In particular, are we trying to provide no-source-change compatibility for internal programs on a best-effort basis (is it part of the compatibility.h mandate?), or are we recommending that they update their code themselves to follow the runtime internal evolution?
- What is the recommended way for user programs to write code that works with several versions of the OCaml runtime?

My impression is that compatibility.h is trying to do too many things at once, and that it should be split in several independent files that provide different services.

**Comment by gasche** [comment]
(cc @damiendoligez @gadmm @jhjourdan @kit-ty-kate @nojb @Octachron @stedolan @xavierleroy)

**Comment by gasche** [comment]
Note: if we cannot find a satisfying resolution for these issues in time for the 4.10 release schedule, I think that we should revert #8713 (Caml_state introduction) from the release branch.

**Comment by gadmm** [comment]
Thanks for the post. I agree with the need for clarification and the current lack of documentation, but note that #9167 does aim to bring some clarification, and does contain documentation changes. I'll try to reply to your questions as soon as I have the time.

**Comment by XVilka** [comment]
I think the "compiling with CompCert" issue is probably also related https://discuss.ocaml.org/t/compiling-the-ocaml-runtime-with-compcert/4898/7

**Comment by gasche** [comment]
No, I don't think the two issues are related. (The Compcert discussion mentioned autoconf as a possible pain point, which is completely different.)

**Comment by xavierleroy** [comment]
Indeed, let's keep CompCert out of this discussion, it's not related.

**Comment by xavierleroy** [comment]
I'm going to give my personal view on the issues.  I think we've followed wrong directions and need to backtrack and make it right this time. ( I know this is disappointing for the many others who already worked on these issues, but when you hit the ground, it's a good idea to stop digging.)

My first and main point is that the 
```
#define caml_foo (Caml_state_field(foo)
```
macro definitions should NOT be in `<caml/compatibility.h>`.  This include file should only deal with the "compatibility with older, short names" business, i.e. `#define foo caml_foo`.

Instead, the `#define caml_foo (Caml_state_field(foo)` should be placed in the header file that used to declare `caml_foo`, at exactly the  same location in this file, so that the definition is guarded by `#ifdef CAML_INTERNALS` if it happens that the declaration was guarded by `CAML_INTERNALS`.

Concrete example: in `<caml/backtrace.h>`, line 58, we had
```
CAMLextern int caml_backtrace_active;
```
This should be replaced by
```
#define caml_backtrace_active (Caml_state_field(backtrace_active))
```
and will therefore be guarded by the `#ifdef CAML_INTERNALS` on line 19.

**Comment by xavierleroy** [comment]
My second point is that short names (e.g. `alloc` instead of `caml_alloc`) must go.  The proper replacements (`caml_alloc`, etc) have been around for a very long time, and have been documented for just as long, while the short names have been undocumented for just as long too.  

I think now -- that is, the introduction of `Caml_state` -- is the right time to get rid of the short names and of `<caml/compatibility.h>` entirely, because 
1. this would avoid the conflicts between `foo` as a short name and `foo` as a member of `struct caml_domain_state`, and
2. release 4.10 is breaking some other things in the C FFI, so, while we are fixing broken packages, let's modernize them too.

**Comment by xavierleroy** [comment]
My third point is that we should make all efforts to maintain compatibility with what's documented in the FFI chapter of the reference manual, and feel free to break everything else.  

Libraries and systems that MUST go beyond what's documented are very few, and can be handled specially.  For example, delimcc MUST go into undocumented land because it tweaks the internals of the bytecode interpreter, but I'm willing to work with Oleg to add the couple of `#if OCAML_VERSION >= 41000` that might be necessary.  For another example, I am not persuaded  that Coq MUST go into undocumented land in its VM, but if it must that's also something we can take care collectively.

**Comment by dra27** [comment]
I agree that life will be much simpler if we drop the short names mode of compatibility.h - we put a lot of effort into trying to keep it with #8713, but it’s clearly not been successful.

There is a caveat with putting the definition in their headers - the original idea was that the compiler itself would never have the compatibility definitions (in order to ensure that our own code didn’t accidentally rely on them) - so we might want to consider an additional variable which should absolutely only be defined by the compiler build (yuck) or using the suggestion in #9167 to prevent the compiler from ever seeing them

**Comment by dra27** [comment]
(Apart from the sorrow it will create, I’m not sure that reverting #8713 will itself be easy, so let’s definitely fix this!)

**Comment by gasche** [comment]
Reverting #8713 would be a pain, but it's "just" a lot of work, there are little unknowns and relatively little impact on users outside the compiler distribution. (And it wouldn't be reverted on trunk, so we would have more time to prepare for a better compatibility story.) I will push for it if we have the impression that fixing the C-compatibility issues is cause for undue stress during the release process, or would delay the release too much (it is currently the reason why we don't have a beta-release out yet).

**Comment by gadmm** [comment]
> I would like to understand answers to the following questions:

Excellent questions, I'll try to answer based on my experience of looking at code and with the rationale I reconstructed. (I do not pretend this was the original intention of multicore developers wrt compatibility.h.)

> 
> What is the responsibility of `compatibility.h`?

Defining backwards-compatibility helpers to simplify the writing of code that is compatible across several versions of OCaml. (I do not assume that this is only for internals.)

To me, it has to be thought in combination with some documentation to help transitioning code. (The documentation does not exist yet, and, again, it would be nice to have one, based on real-world examples.)

For instance, for the introduction of `Caml_state`, depending on the choices you make, users can have anything between:
* many errors messages and repetitive fixup tasks, where they have to sprinkle the code with `#if OCAML_VERSION` or create their own version of `compatibility.h`.
* few error messages with fixes that are regular enough that they can be documented.

Documenting the transition steps is simpler (or even possible) if the typical fixes are simpler. With `compatibility.h` and the proposed PRs, we are closer to the second case. We know that there is the for instance typical external value declaration they have to enclose in `#if OCAML_VERSION` that is enough in some (idk if most) cases, and delimcc for instance should even work out of the box.

> (Who is in charge of the decision to include it in a C program?)

When compatibility.h was introduced: automatically included for the user. When it evolved, it seemed that this question was not addressed. #9167 wants to fix that: like at the beginning, it is included automatically for the user, and controlled with `CAML_NAME_SPACE`.

Xavier suggests to split the definitions and place them in relevant headers. I like this idea, if it can be made to work (this is not clear to me). This is a different way to fix the first part of #9167, and as David says it is orthogonal to how users can turn off these helpers, which will be important in both designs.

> How does a program signal that they use the internals of the OCaml runtime?

They define CAML_INTERNALS. (edit: well, they can also declare `external...`) (edit2: and include one of those undocumented headers recently added to the `caml/` dir...)

> What are our recommendation for upgrading those programs as the runtime evolves?

Cf thoughts on documentation above.


> In particular, are we trying to provide no-source-change compatibility for internal programs on a best-effort basis (is it part of the compatibility.h mandate?), or are we recommending that they update their code themselves to follow the runtime internal evolution?

Not sure about "best-effort" as this does not mean anything clear to me. For sure, do not make it worse than necessary (you could always break all programs that use internals at every release if you wanted...). On the other hand, about internals, you do not want to overthink it.

Think about: "do all users have to reuse the same piece of #ifdef'ed code?". Then it deserves a backwards-compatibility helper. "Does a piece of code allow me to give simpler and more uniform steps when documenting transition?" Then it deserves a backwards-compatibility helper. 


>  What is the recommended way for user programs to write code that works with several versions of the OCaml runtime?

For people who evolve OCaml: the first rule is, do not break code. For code that relies on undocumented features or properties: be on the lookout for legitimate uses, that ought to be documented/better supported. If the use is legitimate, do you leave a clear evolution path? For instance, what we did with Jacques-Henri for the Coq VM denoted a legitimate and natural use that is now officially supported (polling at safe points in long-running C code).

Again, big changes are coming with multicore so I share your sentiment that the evolution should be better documented.

> My impression is that compatibility.h is trying to do too many things at once, and that it should be split in several independent files that provide different services.

To summarise, I have explained how compatibility.h in the sense of #9167 tries to do one thing and in the simplest way.

**Comment by gadmm** [comment]
@xavierleroy

About moving to relevant headers:

* It seemed to me that it was more complicated than that because users sometimes have to rely on making `extern` declarations themselves. I did not try, and I may be wrong.
* `compatibility.h` was already augmented with definitions other than short name macros, the `Caml_state` patch did not invent anything new here.
* One needs a way to disable compatibility helpers, and this comes with some boilerplate.

I do not say it is a bad idea, but the advantages are not as obvious as it seems. I'll be happy to help a PR in this direction with some of my review time.

**Comment by xavierleroy** [comment]
To prevent any misunderstanding: I'm definitely not suggesting we revert  #8713 (the introduction of the `Caml_state` structure), it's a step in the right direction.  I am wondering whether we should continue hacking  `<caml/compatibility.h>` or find another approach to providing old names for the members of `Caml_state`.

> It seemed to me that it was more complicated than that because users sometimes have to rely on making extern declarations themselves. I did not try, and I may be wrong.

There was one example with Coq doing `extern char * caml_young_start;` or something similar.  There is no way this can keep working.  If they really must access `caml_young_start`, they should include `<caml/minor_gc.h>`.

> compatibility.h was already augmented with definitions other than short name macros, 

You're right, for two names:
```
/* internal global variables renamed between 4.02.1 and 4.03.0 */
#define caml_stat_top_heap_size Bsize_wsize(caml_stat_top_heap_wsz)
#define caml_stat_heap_size Bsize_wsize(caml_stat_heap_wsz)
```
That feels strange to me, but I'll ask the original author.

> One needs a way to disable compatibility helpers, and this comes with some boilerplate.

> the original idea was that the compiler itself would never have the compatibility definitions (in order to ensure that our own code didn’t accidentally rely on them) 

I assume we are talking about the `caml_foo` defines, not the `foo` obsolete short names.  How strong is this need to hide the `caml_foo` defines?  From the standpoint of FFI users, these are names in the `caml_` namespace, so no serious name pollution occurs.  From the standpoint of the runtime system itself, is it a) not nice, or b) a catastrophic failure to use one of the old names?

**Comment by dra27** [comment]
For the internal use of the compatibility, it was more that it’s not nice (and including code which uses it _might_ mask bugs). So it could be fixed for 4.10 and guarded properly on trunk later (which would mean no need to hastily merge #9167).

**Comment by dra27** [comment]
My impression is that #9202 gets the compatibility story to the best place it can be for packages (@kit-ty-kate - is that right?) and the  there are some places which will have to have changes (eg the example in Coq). Do we just need to go through and protect those new definitions in `compatibility.h` with `#ifdef CAML_INTERNALS` as appropriate and we’re there for a beta? Then, for 4.11, let’s:

- merge #9167 (in some form)
- make `CAML_NAME_SPACE` the default (and remove the definitions in `compatibility.h`)
- move the remaining `Caml_state` and other definitions to the headers where they were originally defined

`compatibility.h` disappears and the documentation needed is solely the new way `CAML_NAME_SPACE` works and the existing stuff about `CAML_INTERNALS`, right?

**Comment by xavierleroy** [comment]
Grepping through OPAM sources, I found this gem:
```
lablgtk.2.18.8/src/wrappers.h:CAMLextern char *young_start, *young_end; /* from minor_gc.h */
```
Old names + explicit import, tsss....  (Cc @garrigue)

**Comment by xavierleroy** [comment]
Grepping through OPAM sources again, I estimate there are about 200 packages that use old, short names, out of 2528 packages analyzed.  (I looked only at the latest version of each package.)

**Comment by xavierleroy** [comment]
> My impression is that #9202 gets the compatibility story to the best place it can be for packages (@kit-ty-kate - is that right?) and the there are some places which will have to have changes (eg the example in Coq). 

Agreed. I just merged #9202 .

> Do we just need to go through and protect those new definitions in compatibility.h with #ifdef CAML_INTERNALS as appropriate and we’re there for a beta? 

Maybe I misunderstood, but there was also a question of `<caml/compatibility.h>` not being consistently included, causing some of the global names that used to be declared in another header file to disappear.   @gadmm knows better.

**Comment by gadmm** [comment]
A few notes:
* I am not aware of remaining issues with short names. Also, removing short names affects programs that use the documented C FFI, whereas Caml_state mainly affects programs that use internals. This explains the disproportion in affected packages. So, while I proposed to remove short names a while back, I no longer think it is a good opportunity to do so. However:
  * Users should not use the `Caml_state_field` macro, but upgrade their code not to use short names---is it possible to remove it/make it invisible to users? If you enforce the invariant that `CAML_NAME_SPACE` is defined whenever `CAML_INTERNALS` is defined, then code simplifies and few occurrences of `Caml_state_field` remains (the definition of `Is_young` in `address_class.h`, and the occurrences of local_roots in `memory.h`).
  * Separately from this issue, you could decide that it is time to deprecate short names by issuing a compiler warning if compatibility.h or config.h is included without `CAML_NAME_SPACE` being defined. (No personal opinion.)
* Xavier argues that if we consider just this change, we do not necessarily need a switch to turn of the macros. However:
  * the idea is that the mechanism will be useful later (multicore API in #1003, deprecate old identifiers like `Begin/End_roots` in #1798).
  * it is not only for the compiler, but also for users: who might want to clean-up their code later, or simply review what is affected by the change in 4.10.
  * if the choice is between having macros definitions in their "right place" but we cannot turn them off, and having them in the same place and they can be turned off, the latter seems preferable to me. (But you can have `CAML_NAME_SPACE` in both cases.)
* The details of the `CAML_NAME_SPACE` mechanism can indeed be finalised in trunk, and is not absolutely necessary for 4.10. However the general direction should be agreed upon for 4.10 because it affects the rest.

I think the following should lead to an OK state (after that, there are diminishing returns of spending more time on it).

- [x] Merge #9202 
- [ ] Be consistent on when/where the compatibility macros are defined (either with #9167 or with Xavier's proposal to find an appropriate specific header for each macro.)
- [ ] Make `Caml_state` internal back again (#9179).
- [ ] Remove/make invisible `Caml_state_field`.
- [ ] Document transition steps somewhere.

 (Please make sure you like the state before the beta, no change after it unless to fix oversights.)

**Comment by gadmm** [comment]
> Grepping through OPAM sources, I found this gem:
> 
> ```
> lablgtk.2.18.8/src/wrappers.h:CAMLextern char *young_start, *young_end; /* from minor_gc.h */
> ```
> 
> Old names + explicit import, tsss.... (Cc @garrigue)

Xavier, users can do `#include <caml/minor_gc.h>` only since 2016 (#656). Surely you are not expecting users to proactively fix what is not broken.

By the way, maybe this is the real reason why the "each macros in the correct header approach" does not work, it is only backwards compatible up to 4.04. (I do not know whether this is considered old or not.)

**Comment by xavierleroy** [comment]
Apologies for the sarcastic tone of my note, I known I shouldn't have.  But I despair at the quantity of key OCaml libraries that depend on undocumented  internals of the runtime system like it was 15 years ago.

> Surely you are not expecting users to proactively fix what is not broken.

I am.  Deprecation warnings are a way to nudge users in this direction.  Either user code is preventively maintained, or it will break badly when OCaml evolves.  Surely you are not expecting OCaml to preserve backward compatibility forever?

**Comment by gasche** [comment]
I think we could be a bit more proactive adding proper APIs for some of the things users do with internal access if (1) it's reasonable and (2) users are kind enough to let us know what they are doing. For example we could add API support for clearing the backtrace, which is what broke with base in 4.10. I am planning to work on this specific one eventually, but I am busy with TRMC right now.

**Comment by kit-ty-kate** [comment]
I'm not sure how relevant this is but here is something that may be related in ocurl: https://github.com/ygrek/ocurl/commit/967609e2e7861321bded926308a98011b787eadf

Before this commit `ocurl` couldn't compile with OCaml 4.10 but compiled fine with previous versions. The issue is here that `CAML_NAME_SPACE` is not defined when `caml/config.h` is included so the short names get included, then `CAML_NAME_SPACE` is defined and more caml headers are included which have macros like this for instance `#define CAMLdrop Caml_state_field(local_roots) = caml__frame`, which fails because `local_roots` has been defined during the first pass as `#define local_roots caml_local_roots`.

**Comment by xavierleroy** [comment]
> The issue is here that CAML_NAME_SPACE is not defined when caml/config.h is included so the short names get included, then CAML_NAME_SPACE is defined and more caml headers are included

"Doctor, it hurts when I do this!
-- Then don't do it."

`CAML_NAME_SPACE` is intended to be defined on the command-line of the C compiler.  You can define it in source files but then, yes, it should be defined before any includes.

**Comment by gadmm** [comment]
We have an ongoing discussion about the meaning of backwards-compatibility, but to me the current issue was that it was already agreed that there would be some compatibility helpers (and we saw that they help to some degree), but that there are issues with a design that is not finalised and an implementation that is inconsistent. As far as `Caml_state` is concerned, I have nothing more to add to the discussion. Several viable paths have been proposed (including Gabriel's which is not outlandish to me), and now it remains to take an executive decision about the general direction.

Gabriel's questions are reaching a bit further and also concern general backwards-compatibility principles. But I am not sure that the examples brought with the `Caml_state` change are the most relevant ones to reflect about this since the fixes are simple and repetitive and the errors do not teach us anything about how we program with OCaml.

But since Xavier asks, and it somewhat fits Gabriel's questions, I propose two extreme examples. First have a look at the switch from Python 2 to Python 3, decided from the top by changing many details in the name of _the right thing_ and to get rid of "cruft" (including a change to the semantics of strings whose consequences were miscalculated), to see what happens with broad changes requiring a stop-the-world effort when users do not see the benefit. I raise my eyebrows when I hear that evolving OCaml somehow consists in removing cruft by deprecating `Pervasives` or `Array.create` or removing `Dynlink.init`. (Hence my middle-ground position regarding short names.) At the other extreme, have a look at the introduction of move semantics in C++11, originally meant to address a pervasive pattern of inefficiency with deep copies arising in practice, but which turned out to radically change the face of the language, without breaking programs. Most of the design effort went into ensuring that the feature is fully integrated into the language and library, which had the consequence of _not_ splitting the language into incompatible dialects, permitting the progressive migration of legacy code, making a wide range of unsafe practices obsolete (effectively) and driving further evolution of the language. The supposed opposition between remaining backwards-compatible and evolving a language is a false one.

**Comment by Octachron** [comment]
Discussing with @kit-ty-kate , it seems that most of the downstream compatibility issues in opam have now a resolution path.

If I understand correctly, we still have the issue of exposing internals with compatibility macros that should be solved before the full release. But it seems to me that we can proceed with the 4.10 beta.

Rereading the discussion, I have a very naive question: Are we planning to deprecate the
```
#define caml_foo (Caml_state_field(foo))
```
macros at some point in the foreseeable  future? If this is not the case, the inclusion of those macros in compatibility.h does not seem ideal?

**Comment by gadmm** [comment]
Looks good overall, seems like a lot of work. The case of `delimcc` caught my attention though: the proposed patch is to include the header `compatibility.h` by hand. However, none of the solutions currently proposed in this thread involves the user having to include `compatibility.h` by hand at any point. I recommend against the inclusion of this change downstream. Instead, the corresponding bug in the OCaml API should be fixed. Remember that fixing some publicly-available programs is only hiding symptoms, and also that even those packages will hit the same issue when they will start passing `-DCAML_NAME_SPACE` as recommended.

I am also worried that patches are sent downstream while the design of the API has not been finalised. While they seem compatible with my own proposal, this might force some backtracking for other proposals (such as Gabriel's).

Regarding deprecating the `caml_foo` macros, this is not a naive question, and was part of the discussion, in the sense that if one wants to let users undefine these macros (a fortiori open a way to deprecate), then it is a bit easier to have them in `compatibility.h`. If not, it becomes easier to implement @xavierleroy's proposition. However, your question is a reminder that it is still not clear what has been decided, so I do not see how one could even start working towards a solution (I personally exclude helping further with PRs while I am in the dark).

For short names, one idea if you still want to avoid the complications is to only remove the macros that interfere with `Caml_state` (but beware of `local_roots` and `compare_unordered` which are not technically internal, even if they might be considered internal in spirit).

**Comment by Octachron** [comment]
I have tested @xavierleroy proposal to define the `caml_*` macros inside the original header files ([here](https://github.com/Octachron/ocaml/commit/3de9b6d18802dc2dace8a1bc1e67d083e764d84b)). And looking at the package in @kit-ty-kate alpha repository that were patched for 4.10, this path requires a sligthly different patch set. However,  the patches looks more natural to me. For instance, a minial change for coq would be a patch
```diff
< CAMLextern char *young_ptr;
< CAMLextern char *young_limit;
> #include <caml/minor_gc.h>
```
whereas the current patch is more akin to
```diff
> #if OCAML_VERSION < 41000
 CAMLextern char *young_ptr;
 CAMLextern char *young_limit;
> #endif
```
In particular, the fact that in the current state of `compatibility.h`, *all* compatibility macros suddenly pop up in the scope in 4.10 seems non-ideal?

With xavier's proposal, most changes for packages that redefined themselves internal variables are similar to the one above. (with some addition of `#define CAML_INTERNALS` when needed ).

Contrarily, `delimcc` which already imports internal variables by including the right headers after defining `CAML_INTERNALS` compiles without a change with this proposals.

**Comment by gadmm** [comment]
> I have tested @xavierleroy proposal to define the `caml_*` macros inside the original header files ([here](https://github.com/Octachron/ocaml/commit/3de9b6d18802dc2dace8a1bc1e67d083e764d84b)).

Thanks for the initiative. I get that you confirm that you found a canonical place for each definition in some appropriate `caml/` header. On my checklist I also have to ask whether it is ok that the "nice" upgrade path is only backwards-compatible up to 4.04 (you tell me, I have no experience about which releases are considered "old" or not).

> In particular, the fact that in the current state of `compatibility.h`, _all_ compatibility macros suddenly pop up in the scope in 4.10 seems non-ideal?

On the other hand the point is that the user could request that they are not defined. With your approach it makes sense to have the definitions permanent. But I also want to mention that it is still possible to make them conditional (later) with your approach: just make sure to group them (unlike what you do in `backtrace.h`); then some of the boilerplate can be put in `config.h` instead. The only drawback would be to have these conditional definitions spread out instead of grouped in `compatibliity.h`.

> 
> With xavier's proposal, most changes for packages that redefined themselves internal variables are similar to the one above. (with some addition of `#define CAML_INTERNALS` when needed ).

LGTM. (Coq itself will need a more specific patch that you say for other reasons, but indeed what you show is what I expected in the general case.)

> 
> Contrarily, `delimcc` which already imports internal variables by including the right headers after defining `CAML_INTERNALS` compiles without a change with this proposals.

LGTM.

Overall only the 4.04 backwards limit worries me, but you may be able to reassure me on that point.

**Comment by gasche** [comment]
@gadmm was pointing out to me earlier today that we needed someone to make a decision (and, it seems, @xavierleroy was happy to suggest but hasn't been willing to make a decision so far). I propose to decide that @Octachron's approach is a good one, and ask for a PR that implements it. Then we review it, we merge it and we are partly done.

Partly done:
- we still need to ensure that the resulting state of affairs is properly documented in the manual (@Octachron, could you do this?)
- It probably makes sense to document the domain state as well in the manual on the C/FFI interface, if we intend that "modern" code will use it -- it's a rather useful abstraction to think about runtime state.
- I think that `OCAML_VERSION` itself, and its friendly neighbors, should be documented in the manual as well (is someone willing to do this?)

**Comment by kit-ty-kate** [comment]
For what it's worth, from my point of view, I much prefer @gadmm's approach over @Octachron's.
It allows to give a real status to compatibility and should not break any existing packages. For instance, @Octachron's approach breaks `lablgtk` which has already been released and I suspect it also breaks other packages. From my point of view it's already too late to make such a big change. Also I feel like scattering compatibility macros everywhere is not the right solution but I'm not an OCaml compiler dev so I may not have the right hunch for this kind of thing.

**Comment by xavierleroy** [comment]
I am not in a position to make a decision,  as I don't fully understand all the implications and have no time right now to get a deeper understanding.  I thank @Octachron for the implementation of the approach I suggested, and still feel that this approach is somewhat natural (anywhere a global variable was declared there is now a macro), but I defer to @kit-ty-kate extensive experience with repairing broken packages :-)

**Comment by garrigue** [comment]
For practical reasons, the fix in lablgtk was introduced early in the release cycle, but it is unfortunately no more than a hack to work around a moving situation. I have no problem with doing one more release to clean up things.
Looking at the headers, in fact using macros in address_class.h would be the best for lablgtk, but I somehow overlooked it when it was introduced.

**Comment by Octachron** [comment]
>  On my checklist I also have to ask whether it is ok that the "nice" upgrade path is only backwards-compatible up to 4.04 (you tell me, I have no experience about which releases are considered "old" or not).

Concerning compatibility, this is a little more subtle than that: only internal variables are affected, and by that I mean the ones that did not exist at all in any installed headers prior to 4.04 .

For instance, the variables defined in `minor_gc` are not in this case, and they can be included at least for any version of OCaml ≥ 3.08  (which is the oldest opam-avaible version that compiles on my system).

So the only packages affected would be the ones that redefine internal variables to access them, and support ocaml version across all the 4.04.0 to 4.10.0 range.

There were no such packages in the opam repository:

* Coq, ocamnet, lablgtk only use 'minor_gc.h''s variables
* core_kernel access the internal gc_ctrl's variable but is only compatible with OCaml ≥ 4.08

Moreover those packages (if they exists) can use the OCAML_VERSION to reimplement themselves a compatibility layer with a break point at any version between [4.04, 4.10]. So they would have a upgrade path, it would just not be a nice one (and as far as can see for this class of packages, a conditional macro is necessary).

One point of view that might be missing before arbitraging between the two options is the opinion of the multicore team: @kayceesrk , are there any plans to touch the `caml_*` macros defined in `compatibility.h` in the multicore future? Or is it conceivable that they stay in the headers as a shorthand for accessing some fields of the domain state?

**Comment by gadmm** [comment]
> For instance, the variables defined in `minor_gc` are not in this case, and they can be included at least for any version of OCaml ≥ 3.08 (which is the oldest opam-avaible version that compiles on my system).

Are you sure of that? I took https://github.com/ocaml/ocaml/pull/656/files#diff-a617ad63c0e83e969568e6a2dd52aba1 to mean that minor_gc.h was not installed prior to 4.04. Either way, it is possible to judge it is not a big deal.

> One point of view that might be missing before arbitraging between the two options is the opinion of the multicore team: @kayceesrk , are there any plans to touch the `caml_*` macros defined in `compatibility.h` in the multicore future? Or is it conceivable that they stay in the headers as a shorthand for accessing some fields of the domain state?

Indeed: does the `compatibility.h` approach looks good for the future multicore needs? (cc @stedolan) Otherwise, if it does not fit a bigger picture, I agree with Xavier that his solution looks more natural.

Anyway, thanks for all that, it looks very tedious, but this effort looks necessary and will avoid future pain for the users.

**Comment by garrigue** [comment]
Oops, I tried to include `address_class.h`, and I get lots of errors for versions up to 4.09.0, so this doesn't seem to be a good replacement...
Including `minor_gc.h` does indeed work, so I'm quite neutral about which approach to use, if I can be sure enough that it is going to be stable in the future.

**Comment by Octachron** [comment]
@gadmm , after retesting, I was wrong and mislead by some lingering header files on my system. So, indeed the compaitbility issue appears for all packages with set of supported versions larger than `[4.04.0, 4.10.0]`. In the list of opam package tested above, only ocamlnet is impacted. (coq and lablgtk support starts at 4.05.0).

**Comment by gasche** [comment]
I think that it is reasonable to have a solution that is only backward-compatible up to 4.04, and ask that packages that both (1) support older releases and (2) use runtime internals use `OCAML_VERSION` to have a compatible implementation (... especially if `OCAML_VERSION` is properly documented somewhere).

**Comment by kayceesrk** [comment]
> One point of view that might be missing before arbitraging between the two options is the opinion of the multicore team: @kayceesrk , are there any plans to touch the caml_* macros defined in compatibility.h in the multicore future? Or is it conceivable that they stay in the headers as a shorthand for accessing some fields of the domain state?

While there are no plans to touch the `caml_*` macros defined in `compatibility.h`, I agree with the original observations made by Xavier that adding macros in `compatibility.h` wasn't the right decision. I am in favor of Xavier/@Octachron's approach of defining compatibility macros in the respective files. 

Apologies for the backwards compatibility pain caused by #8713. @Octachron I would be happy to help with this effort. I would also be happy to contribute to the list of things mentioned by @gasche in https://github.com/ocaml/ocaml/issues/9205#issuecomment-574850796.

**Comment by Octachron** [comment]
So it is seems that a good path forward is:

- document the OCAML_VERSION macro family and the domain state
- check the state of the opam ecosystem with my branch.

Hopefully, I should have the result of an opamcheck run on my branch (plus some patches) by Monday, and I shall document the `OCAML_VERSION` macros sometimes next week.

**Comment by xavierleroy** [comment]
> document[...] the domain state

What is there to document?  To me the domain state is private to the run-time system and not part of the FFI.  Well, today there is the `caml_compare_unordered` former global variable / current domain state member that can legitimately be accessed as part of the FFI (for custom compare functions), but eventually we should offer a better interface, if only a function to set this variable/field.

**Comment by Octachron** [comment]
My idea was to document the fact that domain state is a private part of the run time, and that every external users using it directly should be prepared to see their code broken on every release.
This is quite similar to the existing documentation for internals.

Similarly, we could document that explicit imports are expected to be brittle, and should be replaced by the inclusion of the right header whenever possible.

**Comment by xavierleroy** [comment]
> every external users using it directly should be prepared to see their code broken on every release

OK!  That's a good message to send :-)

**Comment by damiendoligez** [comment]
@Octachron: do we want to keep this PR open?

**Comment by Octachron** [comment]
I don't think so: the core issue has been solved. If there is any feature wishes or open questions buried in the discussion, it seems better to give them their own issue.
